cli/internal/state/state.go

```diff
@@ -186,6 +206,343 @@ func (s *State) Merge(other *State) (*State, error) {
+		validation.EmptySlice(s.Infrastructure.InitSecret).
+			WithFieldTrace(s, &s.Infrastructure.InitSecret),
+		// ClusterValues must be empty.
+		// As the clusterValues struct contains slices, we cannot use the
+		// Empty constraint on the entire struct. Instead, we need to check
+		// each field individually.
+		validation.Empty(s.ClusterValues.ClusterID).
+			WithFieldTrace(s, &s.ClusterValues.ClusterID),
+		validation.Empty(s.ClusterValues.OwnerID).
+			WithFieldTrace(s, &s.ClusterValues.OwnerID),
+		validation.EmptySlice(s.ClusterValues.MeasurementSalt).
+			WithFieldTrace(s, &s.ClusterValues.MeasurementSalt),
+	}
+}
+
+// preInitConstraints are the constraints on the state that should be enforced
+// *before* a Constellation cluster is initialized. (i.e. before "constellation apply" is run.)
+//
+// The constraints check if the infrastructure state is valid, and if the cluster values
+// are empty, which is required for the cluster to initialize correctly.
+func (s *State) preInitConstraints() []*validation.Constraint {
+	return []*validation.Constraint{
+		// state version needs to be accepted by the parsing CLI.
+		validation.OneOf(s.Version, []string{Version1}).
+			WithFieldTrace(s, &s.Version),
+		// infrastructure must be valid.
+		// out-of-cluster endpoint needs to be a valid DNS name or IP address.
+		validation.Or(
+			validation.DNSName(s.Infrastructure.ClusterEndpoint).
+				WithFieldTrace(s, &s.Infrastructure.ClusterEndpoint),
+			validation.IPAddress(s.Infrastructure.ClusterEndpoint).
+				WithFieldTrace(s, &s.Infrastructure.ClusterEndpoint),
+		),
+		// in-cluster endpoint needs to be a valid DNS name or IP address.
+		validation.Or(
+			validation.DNSName(s.Infrastructure.InClusterEndpoint).
+				WithFieldTrace(s, &s.Infrastructure.InClusterEndpoint),
+			validation.IPAddress(s.Infrastructure.InClusterEndpoint).
+				WithFieldTrace(s, &s.Infrastructure.InClusterEndpoint),
+		),
+		// Node IP Cidr needs to be a valid CIDR range.
+		validation.CIDR(s.Infrastructure.IPCidrNode).
+			WithFieldTrace(s, &s.Infrastructure.IPCidrNode),
+		// UID needs to be filled.
+		validation.NotEmpty(s.Infrastructure.UID).
+			WithFieldTrace(s, &s.Infrastructure.UID),
+		// Name needs to be filled.
+		validation.NotEmpty(s.Infrastructure.Name).
+			WithFieldTrace(s, &s.Infrastructure.Name),
+		// GCP values need to be nil, empty, or valid.
+		validation.Or(
+			validation.Or(
+				// nil.
+				validation.Equal(s.Infrastructure.GCP, nil).
+					WithFieldTrace(s, &s.Infrastructure.GCP),
+				// empty.
+				validation.IfNotNil(
+					s.Infrastructure.GCP,
+					func() *validation.Constraint {
+						return validation.And(
+							validation.EvaluateAll,
+							validation.Empty(s.Infrastructure.GCP.ProjectID).
+								WithFieldTrace(s, &s.Infrastructure.GCP.ProjectID),
+							validation.Empty(s.Infrastructure.GCP.IPCidrPod).
+								WithFieldTrace(s, &s.Infrastructure.GCP.IPCidrPod),
+						)
+					},
+				),
+			),
```

Comment:
Do you agree that this would be much more readable? 
This solution would rely on closures but it seems good enough to me for the improved ergonomy. What do you think?

```suggestion
			*validation.Constraint {
				Satisfied: func() *TreeError {
					if s.Infrastructure.GCP == nil {
						return nil
					} else {
						return validation.Empty(s.Infrastructure.GCP)
					}
				}
			}
```
This syntax could probably be further shortened through a validation.CustomFn() helper.